### ISSUE TITLE
Restrict cipher suites and force minimum TLS version

### ISSUE THREAD

**Issue by stevvooe** [docker/distribution#280]
Per @diogomonica [suggestion](https://github.com/docker/docker/pull/11477#discussion_r26713424), the TLS config should be restricted to a set of cipher suites and require at least TLS 1.0.

The suggestion is recaptured here:

```go
config := tls.Config{
    MinVersion:               tls.VersionTLS10,
    PreferServerCipherSuites: true,
    CipherSuites: []uint16{
        tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
        tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
        tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
        tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
        tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
        tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
        tls.TLS_RSA_WITH_AES_128_CBC_SHA,
        tls.TLS_RSA_WITH_AES_256_CBC_SHA},
}
```

**Comment by diogomonica** [comment]
Just so we are clear, my recommendation for 1.7 is that we use TLS1.2 = between registry-clients and registry itself. This recommendation was based on the lack of time to test compatibility issues for 1.6.

**Comment by stevvooe** [comment]
@diogomonica Thank you for the clarification. We can probably go ahead with 1.2 and test for compatibility issues.

**Comment by stevvooe** [comment]
More clarification on desired action for this ticket: https://github.com/docker/distribution/issues/279#issuecomment-83715422

Summary:

- TLS 1.0 around for older python clients
- Server config should be similar to that described above (allowing TLS 1.0)
- Clients should force TLS 1.2, as referenced in the comment above

**Comment by stevvooe** [comment]
CC @NathanMcCauley @diogomonica

**Comment by thaJeztah** [comment]
I think @ewindisch and @estesp are also working on a default set, but a slightly different order (on my phone now, so don't have the PR/issues at hand)

**Comment by thaJeztah** [comment]
For reference (and since this is in Docker, I'm not 100% sure these should be the same); these are the PRs I was talking about; https://github.com/docker/docker/pull/13848 and https://github.com/docker/docker/pull/13851

**Comment by aaronlehmann** [comment]
The ciphersuite suggestions look good. I'm wondering if we should prefer ECSDA over RSA for performance reasons, though.

Is it really safe for clients to insist on TLS 1.2? What if a registry is behind a load balancer that doesn't support TLS 1.2?

**Comment by aaronlehmann** [comment]
It looks like the client change would have to happen in docker/docker. Its TLS settings are contained in `pkg/tlsconfig`. Currently, for communication with the registry it has TLS 1.0 set as the minimum version. Weirdly, it's using ServerDefault (which seems to be meant for the Docker daemon) instead of ClientDefault. Note that ClientDefault sets TLS 1.2 as the minimum version.

As I mentioned above, I'm not sure we can make the client insist on TLS 1.2 because of load balancers.

**Comment by NathanMcCauley** [comment]
Unforunately @aaronlehmann is probably right w.r.t. load balancers not being up to 1.2 yet.

**Comment by stevvooe** [comment]
@aaronlehmann Like this:

```
config := tls.Config{
    MinVersion:               tls.VersionTLS10,
    PreferServerCipherSuites: true,
    CipherSuites: []uint16{
        tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
        tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
        tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
        tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
        tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
        tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
        tls.TLS_RSA_WITH_AES_128_CBC_SHA,
        tls.TLS_RSA_WITH_AES_256_CBC_SHA},
}
```

`ECDHE > RSA, ECDSA > RSA, GCM_SHA256 > CBC_SHA`

**Comment by aaronlehmann** [comment]
PR here: https://github.com/docker/distribution/pull/789.

I mostly followed your template, but flipped the `tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA` and `tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA` entries, based on @diogomonica's comment in https://github.com/docker/distribution/issues/279#issuecomment-83715422 about preferring AES-128.

I agree with the general order of precedence. GCM should be used whenever possible, because there are attacks on the CBC mode. Failing that, we prefer ciphersuites that use ECDSA over RSA, and the last resort ciphersuites are the non-ephemeral ones.

**Comment by stevvooe** [comment]
@aaronlehmann SGTM

**Comment by aaronlehmann** [comment]
Fixed with the merge of #789
